File: pnrs-tools-back-java/api/src/main/java/com/pnrstools/api/services/SoapService.java
package com.pnrstools.api.services;

import com.pnrstools.api.config.EndpointsConfig;
import com.pnrstools.api.model.enums.DisruptionType;
import com.pnrstools.api.services.SoapEnvelopeFactory.DisruptionData;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class SoapService {

    private final EndpointsConfig endpointsConfig;
    private final SoapEnvelopeFactory envelopeFactory;
    private final WebClient webClient = WebClient.builder()
            .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(1024 * 1024))
            .build();

    public record SoapStepLog(
            String step,
            String method,
            String url,
            String requestBody,
            Integer status,
            boolean success,
            String error,
            String responseData
    ) {}

    public record DisruptionResult(
            String result,
            List<SoapStepLog> logs
    ) {}

    public DisruptionResult executeDisruptionWithLogs(DisruptionData data) {
        String envelope = envelopeFactory.getEnvelope(data);
        String stepName = "Disrupción " + data.disruptionType() + " - " + data.pnr();

        return callSoapServiceWithLogs(
                stepName,
                endpointsConfig.getSoapDisruption(),
                endpointsConfig.getSoapAction(),
                envelope
        );
    }

    private DisruptionResult callSoapServiceWithLogs(String step, String url, String soapAction, String envelope) {
        SoapStepLog log = new SoapStepLog(step, "POST", url, envelope, null, false, null, null);

        try {
            String response = webClient.post()
                    .uri(url)
                    .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_XML_VALUE + "; charset=utf-8")
                    .header("SOAPAction", soapAction)
                    .bodyValue(envelope)
                    .retrieve()
                    .onStatus(status -> status.is4xxClientError() || status.is5xxServerError(),
                            clientResponse -> clientResponse.bodyToMono(String.class)
                                    .flatMap(body -> Mono.error(new RuntimeException("SOAP error: " + body))))
                    .bodyToMono(String.class)
                    .timeout(Duration.ofSeconds(30))
                    .block();

            log = new SoapStepLog(step, "POST", url, envelope, 200, true, null, response);

            if (response != null && response.contains("<faultstring>")) {
                throw new RuntimeException("SOAP fault: " + response.substring(0, Math.min(500, response.length())));
            }

            return new DisruptionResult(response, List.of(log));
        } catch (Exception e) {
            log = new SoapStepLog(step, "POST", url, envelope, null, false, e.getMessage(), null);
            throw new RuntimeException("Error en llamada SOAP", e);
        }
    }

    public Map<String, Object> callMockWithLogs(String step, String url, Map<String, Object> body) {
        Map<String, Object> result = new HashMap<>();
        result.put("success", false);
        result.put("logs", Collections.emptyList());

        try {
            String response = webClient.post()
                    .uri(url)
                    .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                    .bodyValue(body)
                    .retrieve()
                    .bodyToMono(String.class)
                    .timeout(Duration.ofSeconds(30))
                    .block();

            result.put("success", true);
            result.put("data", response);
            result.put("logs", List.of(Map.of(
                    "step", step,
                    "method", "POST",
                    "url", url,
                    "requestBody", body,
                    "status", 200,
                    "success", true,
                    "responseData", response
            )));
            return result;
        } catch (Exception e) {
            result.put("success", false);
            result.put("message", e.getMessage());
            result.put("logs", List.of(Map.of(
                    "step", step,
                    "method", "POST",
                    "url", url,
                    "requestBody", body,
                    "success", false,
                    "error", e.getMessage()
            )));
            return result;
        }
    }

    public boolean checkVpnStatus() {
        try {
            webClient.head()
                    .uri(endpointsConfig.getMockInt())
                    .retrieve()
                    .toBodilessEntity()
                    .timeout(Duration.ofSeconds(5))
                    .block();
            return true;
        } catch (Exception e) {
            log.warn("VPN check failed: {}", e.getMessage());
            return false;
        }
    }

    public Map<String, Object> callMockIntWithLogs(String pnr, String surname) {
        Map<String, Object> body = new HashMap<>();
        body.put("locator", pnr);
        body.put("surname", surname);
        return callMockWithLogs("Mock INT", endpointsConfig.getMockInt(), body);
    }

    public Map<String, Object> callMockPreWithLogs(String pnr, String surname) {
        Map<String, Object> body = new HashMap<>();
        body.put("locator", pnr);
        body.put("surname", surname);
        return callMockWithLogs("Mock PRE", endpointsConfig.getMockPre(), body);
    }
}

File: pnrs-tools-back/src/services/soap.service.ts
import { ENDPOINTS } from '../config/endpoints.config.js';
import { SoapEnvelopeFactory } from './SOAP/soap-envelope.factory.js';
import type { DisruptionData } from './SOAP/soap-envelope.factory.js';

export interface SoapStepLog {
  step: string;
  method: string;
  url: string;
  requestBody?: string;
  status?: number;
  success: boolean;
  error?: string;
  responseData?: string;
}

// Cliente HTTP base reutilizable para llamadas SOAP
async function callSoapServiceWithLog(
  step: string,
  url: string,
  soapAction: string,
  envelope: string
): Promise<{ result: string; log: SoapStepLog }> {
  const log: SoapStepLog = {
    step,
    method: 'POST',
    url,
    requestBody: envelope,
    success: false
  };

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'text/xml; charset=utf-8',
        'SOAPAction': soapAction
      },
      body: envelope
    });

    log.status = response.status;
    const responseText = await response.text();
    log.responseData = responseText;

    if (!response.ok || responseText.includes('<faultstring>')) {
      throw new Error(`SOAP error: ${responseText.substring(0, 500)}`);
    }

    log.success = true;
    return { result: responseText, log };
  } catch (error: any) {
    log.error = error.message;
    throw { log, originalError: error };
  }
}

// Cliente HTTP base para Mock JSON
async function callMockWithLog(step: string, url: string, body: any): Promise<{ result: any; log: SoapStepLog }> {
  const log: SoapStepLog = { step, method: 'POST', url, requestBody: JSON.stringify(body, null, 2), success: false };
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body)
    });

    log.status = response.status;
    const data = await response.json();
    log.responseData = JSON.stringify(data, null, 2);

    if (!response.ok) throw new Error(`Mock error: ${response.status}`);

    log.success = true;
    return { result: data, log };
  } catch (error: any) {
    log.error = error.message;
    throw { log, originalError: error };
  }
}

export const disruptionService = {
  // Ejecuta Mock INT registrando logs
  async callMockIntWithLogs(pnr: string, surname: string) {
    const body = { locator: pnr, surname };
    return callMockWithLog('Mock INT', ENDPOINTS.MOCK_INT, body);
  },

  // Ejecuta Mock PRE registrando logs
  async callMockPreWithLogs(pnr: string, surname: string) {
    const body = { locator: pnr, surname };
    return callMockWithLog('Mock PRE', ENDPOINTS.MOCK_PRE, body);
  },

  // Envío unificado de Disrupciones (UN, UNTK, FLCH) utilizando Factory Pattern
  async executeDisruptionWithLogs(data: DisruptionData): Promise<{ result: string; logs: SoapStepLog[] }> {
    const envelope = SoapEnvelopeFactory.getEnvelope(data);
    const stepName = `Disrupción ${data.disruptionType} - ${data.pnr}`;

    try {
      const { result, log } = await callSoapServiceWithLog(
        stepName,
        ENDPOINTS.SOAP_DISRUPTION,
        ENDPOINTS.SOAP_ACTION,
        envelope
      );
      return { result, logs: [log] };
    } catch (err: any) {
      if (err.log) return { result: '', logs: [err.log] };
      throw err;
    }
  },

  // Verificación del estado de VPN / Conectividad
  async checkVpnStatus(): Promise<boolean> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5000);
    try {
      await fetch(ENDPOINTS.MOCK_INT, {
        method: 'HEAD',
        signal: controller.signal
      });
      clearTimeout(timeout);
      return true;
    } catch (error) {
      clearTimeout(timeout);
      return false;
    }
  },

  // Wrappers de conveniencia
  async callDisruptionUNWithLogs(data: any) {
    return this.executeDisruptionWithLogs({ ...data, disruptionType: 'UN' });
  },

  async callDisruptionUNTKWithLogs(data: any) {
    return this.executeDisruptionWithLogs({ ...data, disruptionType: 'UNTK' });
  },

  async callDisruptionFLCHWithLogs(data: any) {
    return this.executeDisruptionWithLogs({ ...data, disruptionType: 'FLCH' });
  }

};
